include/swift/Runtime/Mutex.h

```diff
@@ -10,124 +10,137 @@
+/// Use ReadWriteLock or ScopedWriteXxxx instead (see below).
+class WriteMutexImpl {
+  friend class ReadWriteLock;
 
-  MutexImpl(const MutexImpl &) = delete;
-  MutexImpl &operator=(const MutexImpl &) = delete;
-  MutexImpl(MutexImpl &&) = delete;
-  MutexImpl &operator=(MutexImpl &&) = delete;
+  WriteMutexImpl() = delete;
+  WriteMutexImpl(const WriteMutexImpl &) = delete;
+  WriteMutexImpl &operator=(const WriteMutexImpl &) = delete;
+  WriteMutexImpl(WriteMutexImpl &&) = delete;
+  WriteMutexImpl &operator=(WriteMutexImpl &&) = delete;
 
 public:
-  void lock();
-  void unlock();
-  bool try_lock();
-  void wait(Condition &condition);
+  void lock() { Impl.writeLock(); }
+  void unlock() { Impl.unlock(); }
+  bool try_lock() { return Impl.try_writeLock(); }
 
 private:
-  explicit MutexImpl(bool checked);
-
-#ifdef SWIFT_RUNTIME_MUTEX_HAVE_PHTREADS
-  pthread_mutex_t PThreadMutex;
-#endif
+  explicit WriteMutexImpl(ReadWriteLockImpl &impl) : Impl(impl) {}
+  ReadWriteLockImpl &Impl;
 };
 
-/// Internal (private) implementation of scoped locking functionality.
+/// Private implementation of ScopedLock functionality.
 /// Use ScopedLock instead (see below).
-class ScopedLockImpl {
+template <typename MutexType> class ScopedLockImpl {
   friend class Mutex;
+  friend class ReadWriteLock;
   friend class ScopedLock;
+  friend class ScopedReadLock;
+  friend class ScopedWriteLock;
 
-public:
   ScopedLockImpl() = delete;
-  ~ScopedLockImpl() { Impl.unlock(); }
-
   ScopedLockImpl(const ScopedLockImpl &) = delete;
   ScopedLockImpl &operator=(const ScopedLockImpl &) = delete;
   ScopedLockImpl(ScopedLockImpl &&) = delete;
   ScopedLockImpl &operator=(ScopedLockImpl &&) = delete;
 
+public:
+  ~ScopedLockImpl() { Impl.unlock(); }
+
 private:
-  ScopedLockImpl(MutexImpl &impl) : Impl(impl) { Impl.lock(); }
-  MutexImpl &Impl;
+  ScopedLockImpl(MutexType &impl) : Impl(impl) { Impl.lock(); }
+  MutexType &Impl;
 };
 
-/// Internal (private) implementation of scoped unlocking functionality.
+/// Private implementation of ScopedUnlock functionality.
 /// Use ScopedUnlock instead (see below).
-class ScopedUnlockImpl {
+template <typename MutexType> class ScopedUnlockImpl {
   friend class Mutex;
   friend class ScopedUnlock;
+  friend class ScopedReadUnlock;
+  friend class ScopedWriteUnlock;
 
-public:
   ScopedUnlockImpl() = delete;
-  ~ScopedUnlockImpl() { Impl.lock(); }
-
   ScopedUnlockImpl(const ScopedUnlockImpl &) = delete;
   ScopedUnlockImpl &operator=(const ScopedUnlockImpl &) = delete;
   ScopedUnlockImpl(ScopedUnlockImpl &&) = delete;
   ScopedUnlockImpl &operator=(ScopedUnlockImpl &&) = delete;
 
+public:
+  ~ScopedUnlockImpl() { Impl.lock(); }
+
+private:
+  ScopedUnlockImpl(MutexType &impl) : Impl(impl) { Impl.unlock(); }
+  MutexType &Impl;
+};
+
+/// A Condition that works with Mutex to allow – as an example – multi-threaded
+/// producers and consumers to signal each other in a safe way.
+class Condition {
```

Comment:
It may. I followed the pthread naming for now. I can look at the typical semantics for condition variable to ensure it doesn't deviate from the typical to much.